torch/_guards.py

```diff
@@ -175,3 +176,31 @@ def set_export_info(self, guard_type, guarded_class, code_list, obj_weakref):
             None,
         ), "Guarded object must be identical, or None"
         self.obj_weakref = obj_weakref
+
+
+class GuardsContext:
+    def __init__(self):
+        self.dynamo_guards: Set[Guard] = set()
+
+
+_CURRENT_TRACING_CONTEXT = None
+
+
+class TracingContext:
+    @staticmethod
+    def get() -> Optional["TracingContext"]:
+        return _CURRENT_TRACING_CONTEXT
+
+    def __init__(self):
+        self.guards_context = GuardsContext()
+
+
+@contextmanager
+def tracing(context: TracingContext):
```

Comment:
I prefer to keep it separate, as we make it but don't always enter until later - no strong opinions tho.

Reply:
I still prefer my suggestion, you can create the context and only enter when you call `with` on it. These two concepts seem inextricably linked, so I don't see a compelling reason why they should be separate

Reply:
So one oddity in the current PR is that the tracing context is not actually setup for the dynamo context; at the moment it is only setup before going to the user compiler. This.... kind of makes sense, in the sense that the context is mostly useful for user compilers, since we thread the context along in Dynamo proper, but it's kind of tricky.